google/cloud/internal/resumable_async_streaming_read_write_rpc.h

```diff
@@ -0,0 +1,275 @@
+#include <thread>
+
+namespace google {
+namespace cloud {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+namespace internal {
+
+/**
+ * `ResumableAsyncStreamingReadWriteRpc<ResponseType, RequestType>` uses
+ * callables compatible with this `std::function<>` to create new streams.
+ */
+template <typename ResponseType, typename RequestType>
+using AsyncStreamFactory = std::function<
+    std::unique_ptr<AsyncStreamingReadWriteRpc<RequestType, ResponseType>>()>;
+
+/**
+ * `ResumableAsyncStreamingReadWriteRpc<ResponseType, RequestType>` uses
+ * callables compatible with this `std::function<>` to update the request object
+ * after each response. This is how users of the class can update the resume
+ * token or any other parameters needed to restart a stream from the last
+ * received message.
+ */
+template <typename ResponseType, typename RequestType>
+using StreamReinitializer = std::function<future<StatusOr<
+    std::unique_ptr<AsyncStreamingReadWriteRpc<RequestType, ResponseType>>>>(
+    std::unique_ptr<AsyncStreamingReadWriteRpc<RequestType, ResponseType>>)>;
+
+template <typename ResponseType, typename RequestType>
+class ResumableAsyncStreamingReadWriteRpc
+    : public AsyncStreamingReadWriteRpc<RequestType, ResponseType> {
+ public:
+  virtual future<Status> Status() = 0;
+};
+
+template <typename ResponseType, typename RequestType>
+class ResumableAsyncStreamingReadWriteRpcImpl
+    : public ResumableAsyncStreamingReadWriteRpc<RequestType, ResponseType> {
+ public:
+  ResumableAsyncStreamingReadWriteRpcImpl(
+      std::unique_ptr<RetryPolicy const> retry_policy,
+      std::unique_ptr<BackoffPolicy const> backoff_policy,
+      const AsyncSleeper& sleeper,
+      AsyncStreamFactory<ResponseType, RequestType>& stream_factory,
+      StreamReinitializer<ResponseType, RequestType>& reinitializer)
+      : retry_policy_prototype_(std::move(retry_policy)),
+        backoff_policy_prototype_(std::move(backoff_policy)),
+        sleeper_(std::move(sleeper)),
+        stream_factory_(std::move(stream_factory)),
+        reinitializer_(std::move(reinitializer)),
+        impl_(stream_factory_()) {}
+
+  ResumableAsyncStreamingReadWriteRpcImpl(ResumableStreamingReadRpc&&) = delete;
+  ResumableAsyncStreamingReadWriteRpcImpl& operator=(
+      ResumableStreamingReadRpc&&) = delete;
+
+  void Cancel() override {
+    // not guarding since this is the singular writer
+    is_canceled_ = true;
+    std::lock_guard<std::mutex> g{mu_};
+    impl_->Cancel();
+  }
+
+  future<bool> Start() override {
+    std::lock_guard<std::mutex> g{mu_};
+    return impl_->Start().then([this](future<bool> start_fu) {
+      if (!start_fu.get()) {
+        std::shared_ptr<RetryPolicy const> retry_policy =
+            this->retry_policy_prototype_->clone();
+        std::shared_ptr<BackoffPolicy const> backoff_policy backoff_policy =
+            this->backoff_policy_prototype_->clone();
+        return this->RetryLoop(retry_policy, backoff_policy, RetryType.Start);
```

Comment:
you can still abandon this future. Instead of the complex retry_type logic, just write instead for Start:

Restart(Status(StatusCode.kUnavailable, "Starting stream"));
return true;